Refactor cudf::string_view host and device code

davidwendt opened the pull request:
While working on improving the sort performance for strings columns in #7075, we tried a vector-load approach in the `string_view::compare()` function. This approached used some CUDA math intrinsic functions like `__funnelshift_r()` and `__byte_perm()`. Unfortunately, adding these to the `string_view` source would cause compile errors for some .cpp files. This is because the `string_view.cuh` was being included by some .cpp file even though these only used the appropriate `__host__ __device__` functions.

This PR breaks up the host/device from the device-only functions so the .cpp files can include `string_view.cuh` without processing the device-only definitions. The host/device functions are now defined in the `string_view.cuh` directly and the device-only source is isolated in the `string_view.inl`. The include of the `string_view.inl` is then wrapped if a `#if CUDA_ARCH` so it will not be processed by a .cpp file compilation.

Also, I attempted to minimize includes of `string_view.cuh` by removing it from `traits.hpp` and replacing it with a forward reference. This found a few files that were not including `string_view.cuh` directly as they should've. This also exposed `cpp/tests/utilities/scalar_utilities.cu` which appears to be unused and thus removed along with its header.

No functionality has changed. Build times may be slightly faster since `string_view.cuh` is included in less source files and .cpp files no longer the `string_view.inl`. This means changing this file was also have a slightly less impact on rebuilding libcudf.

kkraus14 reviewed (approved):
cmake / conda lgtm

karthikeyann commented on cpp/include/cudf/utilities/traits.hpp:
```diff
@@ -505,6 +504,8 @@ constexpr inline bool is_fixed_width(data_type type)
   return cudf::type_dispatcher(type, is_fixed_width_impl{});
 }
 
+class string_view;
+
```
cudf/types.hpp already has forward declaration of cudf::string_view. This forward declaration may not be needed.

davidwendt replied:
Maybe, but I feel this makes the header file more complete/clear on its own.

karthikeyann commented on cpp/src/strings/regex/regexec.cu:
```diff
@@ -22,6 +22,8 @@
 #include <rmm/device_buffer.hpp>
 
+#include <algorithm>
+
```
copyright year update in this file.

karthikeyann commented on cpp/tests/transform/mask_to_bools_test.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Do you have script or hack to update Copyright year in all files?

davidwendt replied:
No.

karthikeyann reviewed (changes requested):
Neat Refactoring! LGTM. 
Minor corrections.

jrhemstad commented on cpp/include/cudf/strings/string_view.cuh:
```diff
@@ -37,47 +57,28 @@ using char_utf8 = uint32_t;  ///< UTF-8 characters are 1-4 bytes
    * @brief Return the number of bytes in this string
    */
-  __host__ __device__ size_type size_bytes() const;
+  __host__ __device__ inline size_type size_bytes() const { return _bytes; }
```
There's no need to put `inline` on member functions defined inside the class.

jrhemstad commented on cpp/include/cudf/strings/string_view.cuh:
```diff
@@ -37,47 +57,28 @@ using char_utf8 = uint32_t;  ///< UTF-8 characters are 1-4 bytes
    * @brief Return true if string has no characters
    */
-  __host__ __device__ bool empty() const;
+  __host__ __device__ inline bool empty() const { return size_bytes() == 0; }
```
These should use `CUDA_HOST_DEVICE_CALLABLE` so the host/device decorators are removed in `.cpp` files.

jrhemstad commented on cpp/include/cudf/strings/string_view.cuh:
```diff
@@ -322,31 +399,25 @@ __host__ __device__ size_type to_char_utf8(const char* str, char_utf8& character
 
+// Using device specific intrinsic functions in device only code causes a
+// compile error for .cpp files.
+// Many .cpp files need only the host member functions defined above.
```
I'm confused, when does a .cpp file ever need to interact with a `string_view` object?

davidwendt replied:
This is because the string_view.cuh was being included by some .cpp file even though they only need to use __host__ __device__ functions and never the __device__ functions.

jrhemstad replied:
What I'm asking is when does host code ever need to use _any_ of the functions of a `string_view`?

davidwendt replied:
One example is here creating a `string_scalar` from a `string_view`.
https://github.com/rapidsai/cudf/blob/639049808913c7d5a7170141d784da6cc45da36d/cpp/include/cudf/scalar/scalar.hpp#L430-L437
This does not require any device code. Also, getting the `string_view` from a `string_scalar`
https://github.com/rapidsai/cudf/blob/639049808913c7d5a7170141d784da6cc45da36d/cpp/include/cudf/scalar/scalar.hpp#L473-L476
And `scalar.hpp` is included in many .cpp files.
This is just an example of where a `string_view` is safely referenced/used in a .cpp file since it does not require calling any device functions.

jrhemstad replied:
Great, that makes more sense. Thanks for the examples. Those seem valid.

jrhemstad commented on cpp/include/cudf/strings/string_view.cuh:
```diff
@@ -322,31 +399,25 @@ __host__ __device__ size_type to_char_utf8(const char* str, char_utf8& character
+// Many .cpp files need only the host member functions defined above.
+// Therefore, the device only functions are isolated in the .inl file
+// included below.
+#if defined(__CUDA_ARCH__)
```
`__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

davidwendt replied:
I need something like this to prevent including the `.inl` file when compiled by the host compiler and to only include the `.inl` when compiled by nvcc.
Otherwise, I could add it to every `__device__` function inside the `.inl` file instead.

karthikeyann replied:
> `__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

[CUDA Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#application-compatibility
) says `It is only defined for device code`. 
To be specific, it's device compilation phase. `when __CUDA_ARCH__ is defined (i.e., during device compilation phase)` 

`__CUDA_ARCH__` restrictions https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-arch-macro

nvcc can be identified using `__CUDACC__`. https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-identification-macro 
https://stackoverflow.com/a/8809924/1550940

To summarize,
`__CUDA_ARCH__` is defined during device compilation phase (for nvcc host compilation phase, it's not defined).
`__CUDACC__` is defined only when nvcc is compiling CUDA source files. (for C/C++ files it's not defined). But it is defined for both device compilation and host compilation phase.

davidwendt replied:
> __CUDACC__ is defined only when nvcc is compiling CUDA source files
This is exactly what I need.

davidwendt commented:
I decided to take a different approach than using the `__CUDA_ARCH__` define around the `string_view.inl`. 
I've renamed `string_view.cuh` to `string_view.hpp` since it contains only host or host/device code definitions.
I've renamed `string_view.inl` to `string_view.cuh` since it only contains device code.
The `string_view.cuh` includes `string_view.hpp` so if a `.cu` file needs to call any device functions, it only needs to include `string_view.cuh`. All others files can just contain `string_view.hpp`.
This removes the awkward include of a `.cuh` file in a non `.cu` file for example.
It also means we do not need the `__CUDA_ARCH__` if-define logic anymore.
I feel this is much cleaner and easier to understand.

harrism commented:
> No function as changed.

Is this a typo? Having trouble understanding what it means. :)

davidwendt commented:
> > No function as changed.
> 
> Is this a typo? Having trouble understanding what it means. :)

Sorry. I meant "No function _has_ changed." I fixed the description.

jrhemstad reviewed (approved):
Nice solution.

davidwendt commented:
@gpucibot merge
